test/rng_quality.cpp

```diff
@@ -0,0 +1,240 @@
+    expected *= convert<float>(sum<float>(input)) /
+                convert<float>(sum<float>(expected));
+    array diff  = input - expected;
+    double chi2 = convert<float>(sum<float>((diff * diff) / expected));
+    return chi2;
+}
+
+template<typename T>
+void testRandomEngineUniformChi2(randomEngineType type) {
+    SUPPORTED_TYPE_CHECK(T);
+    dtype ty = (dtype)dtype_traits<T>::af_type;
+
+    int elem  = 256 * 1024 * 1024;
+    int steps = 256;
+    int bins  = 100;
+
+    array total_hist = constant(0.0, bins, f32);
+    array expected   = constant(1.0 / bins, bins, f32);
+
+    randomEngine r(type, 0);
+
+    // R> qchisq(c(5e-6, 1 - 5e-6), 99)
+    // [1]  48.68125 173.87456
+    float lower(48.68125);
+    float upper(173.87456);
+
+    bool prev_step  = true;
+    bool prev_total = true;
+    for (int i = 0; i < steps; ++i) {
+        array rn_numbers = randu(elem, ty, r);
+        array step_hist  = histogram(rn_numbers, bins, 0.0, 1.0);
+        step_hist        = step_hist.as(f32);
+        float step_chi2  = chi2_statistic<float>(step_hist, expected);
+        if (!prev_step) {
+            EXPECT_GT(step_chi2, lower) << "at step: " << i;
+            EXPECT_LT(step_chi2, upper) << "at step: " << i;
+        }
+        prev_step = step_chi2 > lower && step_chi2 < upper;
+
+        total_hist += step_hist;
+        float total_chi2 = chi2_statistic<float>(total_hist, expected);
+        if (!prev_total) {
+            EXPECT_GT(total_chi2, lower) << "at step: " << i;
+            EXPECT_LT(total_chi2, upper) << "at step: " << i;
+        }
+        prev_total = total_chi2 > lower && total_chi2 < upper;
+    }
+}
+
+TYPED_TEST(RandomEngine, philoxRandomEngineUniformChi2) {
+    testRandomEngineUniformChi2<TypeParam>(AF_RANDOM_ENGINE_PHILOX_4X32_10);
+}
+
+TYPED_TEST(RandomEngine, threefryRandomEngineUniformChi2) {
+    testRandomEngineUniformChi2<TypeParam>(AF_RANDOM_ENGINE_THREEFRY_2X32_16);
+}
+
+TYPED_TEST(RandomEngine, mersenneRandomEngineUniformChi2) {
+    testRandomEngineUniformChi2<TypeParam>(AF_RANDOM_ENGINE_MERSENNE_GP11213);
+}
+
+// should be used only for x <= 5 (roughly)
+
+array cnd(array x) { return 0.5 * erfc(-x * sqrt(0.5)); }
+
+template<typename T>
+bool testRandomEngineNormalChi2(randomEngineType type)
+
+{
+    af::dtype ty = (af::dtype)af::dtype_traits<T>::af_type;
+
+    int elem  = 256 * 1024 * 1024;
+    int steps = 64;  // 256 * 32;
+    int bins  = 100;
+
+    T lower_edge(-7.0);
+    T upper_edge(7.0);
+
+    array total_hist = af::constant(0.0, 2 * bins, f32);
+    array edges      = af::seq(bins + 1) / bins * lower_edge;
+    array expected   = -af::diff1(cnd(edges));
+
+    expected =
+        af::join(0, expected(af::seq(bins - 1, 0, -1)), expected).as(f32);
+    // af_print(expected);
+
+    af::randomEngine r(type, 0);
+
+    // R> qchisq(c(5e-6, 1 - 5e-6), 197)
+
+    // [1] 121.3197 297.2989
+    float lower(121.3197);
+    float upper(297.2989);
```

Comment:
Nitpick: I would remove the empty line between the comments.

Explanation: In the chi^2 test one computes the test statistic and compares the value with the chi^2 distribution with appropriate number of degrees of freedom. For th euniform distribution one has "number of bins minus 1" degrees of freedom. For the normal distribution it is "number of bins minus 3", since there are two parameters mu and sigma. Here I used the `qchisq()` function from R to compute "suitable" values from the chi^2 distribution.